File: topicos/poo/java/ex2/Biblioteca.java
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

public class Biblioteca {
  private List<Livro> colecao;
  private List<Emprestimo> emprestimos;

  public Biblioteca() {
    this.colecao = new ArrayList<>();
    this.emprestimos = new ArrayList<>();
  }

  public Biblioteca(List<Livro> livros) {
    this.colecao = new ArrayList<>(livros);
    this.emprestimos = new ArrayList<>();
  }

  public void addLivroColecao(Livro livro) {
    this.colecao.add(livro);
  }

  public Emprestimo registraEmprestimo(Livro livro, Leitor leitor) {
    Emprestimo emprestimo = new Emprestimo(livro, leitor, new Date());
    this.emprestimos.add(emprestimo);
    leitor.setLivro(livro);
    return emprestimo;
  }

  public Livro buscarLivro(String autor) {
    for (Livro livro : this.colecao) {
      if (livro.getAutor().equals(autor)) {
        return livro;
      }
    }
    return null;
  }

  private static class Emprestimo {
    private Livro livro;
    private Leitor leitor;
    private Date momento;

    public Emprestimo(Livro livro, Leitor leitor, Date momento) {
      this.livro = livro;
      this.leitor = leitor;
      this.momento = momento;
    }
  }
}

File: topicos/poo/node/ex2/Biblioteca.js
export class Biblioteca {
  #colecao;
  #emprestimos;

  constructor(livros = []) {
    this.#colecao = [...livros];
    this.#emprestimos = [];
  }


  addLivroColecao(livro) {
    this.#colecao.push(livro);
  }

  registraEmprestimo(livro, leitor) {
    const emprestimo = {
      livro,
      leitor,
      momento: new Date()
    };

    this.#emprestimos.push(emprestimo);
    leitor.setLivro(livro);

    return emprestimo;
  }

  buscarLivro(autor) {
    return this.#colecao.find(item => item.autor === autor);
  }
}
